#159965266 Users should be able to login to Author's Haven

mikenthiwa opened the pull request:
#### What does this PR do?
 * Implementation of login feature for front-end.

#### Description of Task to be completed?
 * The login feature allows a user to log in via front-end.

#### How should this be manually tested?
 * Clone our repository: https://github.com/andela/mag-ah-frontend.git
 * Checkout to: ft-login-159965266
 * Install requirements: `npm install`
 * Run app: `npm start`

On the header of the page :
 * Click on Sign in button
 * Provide email and password on the modal that appears

#### What are the relevant pivotal tracker stories?
 * Homepage
<img width="1095" alt="screen shot 2018-10-17 at 11 50 45" src="https://user-images.githubusercontent.com/30254491/47074394-fb23e000-d202-11e8-936d-744137fcf8e8.png">

 * Sign in modal
<img width="616" alt="screen shot 2018-10-17 at 11 52 57" src="https://user-images.githubusercontent.com/30254491/47074538-46d68980-d203-11e8-915d-8968b14d4a79.png">



[User should be able to login to the application](https://www.pivotaltracker.com/n/projects/#159965266)

mikenthiwa committed:
[login #[159965266]] Implement login functionality and design UI

mikenthiwa committed:
[feat(login): #159965266]
 - Implement login functionality
 - Write tests for login component, reducer and action

[Delivers #159965266]

meshack-mbuvi commented:
@mikenthiwa and @Bileonaire please change the color for google sign in button to conform to what PO said.

Bileonaire commented:
Thank you @meshack-mbuvi for your review.
We did not focus on the google button feedback because @pngg is working on the social login and took note of that.

mikenthiwa commented:
@meshack-mbuvi We have resolved the background colour of Google social login button.

kimotho-njoki commented:
First things first, not at all a fan of the Sign In link. Feels too basic. @mikenthiwa @Bileonaire . And once I click on the `signIn` button on the login modal, is it supposed to lead me somewhere, currently it's just looking at me

kimotho-njoki commented on src/__tests__/actions/testLoginAction.js:
```diff
@@ -0,0 +1,46 @@
+  beforeEach(() => moxios.install());
+  afterEach(() => moxios.uninstall());
+
+  it("should dispatch login successful", async done => {
```
`done` is a representation of `()` in this case. Replace with `()` and delete line 26. Please do the same for all scenarios you've used done to avoid adding unnecessary extra lines of code.

kimotho-njoki commented on src/__tests__/actions/testLoginAction.js:
```diff
@@ -0,0 +1,46 @@
+      response: loginSuccefullMessage
+    });
+
+    const store = mockstore({});
```
`store` is being used in multiple places. Define it in `beforeEach()` but then clear all the actions in `afterEach()`, figure out how to do this, this is to ensure that each time a new test runs, the store is reset.

kimotho-njoki commented on src/components/AuthButton.jsx:
```diff
@@ -0,0 +1,73 @@
+
+class AuthButton extends React.Component {
+  constructor(props) {
+    super(props);
```
Are we inheriting from somewhere?

kimotho-njoki commented on src/components/AuthButton.jsx:
```diff
@@ -0,0 +1,73 @@
+  isAuthenticated: PropTypes.bool.isRequired
+};
+
+const mapStateToProps = ({ loginReducer }) => loginReducer;
```
What props are you accessing from `loginReducer`? Specify in such a way

const mapStateToProps = ({ loginReducer }) => {
  const { message, ... } = loginReducer;
  return {
  message,
  ...
  };
};

kimotho-njoki commented on src/components/login.jsx:
```diff
@@ -0,0 +1,77 @@
+    super(props);
+    this.state = { email: "", password: "" };
+    this.handleChange = this.handleChange.bind(this);
+    this.onLogin = this.onLogin.bind(this);
```
I can't recall the last time I saw `binding` statements within a component. I believe if you define `onLogin` and `handleChange` as ES6 statements, you could do away with the above constructor and bind statements, remaining with the state on it's own. Please refer to https://medium.freecodecamp.org/react-binding-patterns-5-approaches-for-handling-this-92c651b5af56 #5 and refactor all components accordingly. Unless you have a preference to binding statements?

kimotho-njoki commented on src/components/AuthButton.jsx:
```diff
@@ -0,0 +1,73 @@
+
+    return (
+      <div>
+        {isAuthenticated ? (
```
Ternary operator arrangement according to Airbnb, can't find a link currently:

`isAuthenticated ? blahblah : blahblah` works for short ternaries

isAuthenticated 
     ?  blahblah
     :   blahblah
For longer ternaries like the one you have now (Please click on edit to see the spacing before the ? and :)

Bileonaire replied:
@kimotho-njoki the auto formatting is formatting it to be the way it is currently. I do not know how to bypass it because It is raising an error if I arrange the ternary differently.

kimotho-njoki commented on src/components/AuthButton.jsx:
```diff
@@ -0,0 +1,73 @@
+                    My profile
+                  </a>
+                  <a href="/" className="dropdown-item" onClick={this.onLogout}>
+                    Logout
```
Where exactly on the page do we have `profile` and `logout`? I cant seem to find the two on the UI

mikenthiwa replied:
@kimotho-njoki: Once a user has logged in, you will be able to see an avatar with a drop down at the top right of the homepage that will have a profile link and log out link. Currently, the profile link is not doing anything since the front end feature is not implemented.
<img width="1019" alt="screen shot 2018-10-15 at 21 28 31" src="https://user-images.githubusercontent.com/30254491/46970307-65346c00-d0c1-11e8-9a50-b5894a566b67.png">

kimotho-njoki commented on src/components/login.jsx:
```diff
@@ -0,0 +1,77 @@
+  error: PropTypes.shape().isRequired
+};
+
+const mapStateToProps = ({ loginReducer }) => loginReducer;
```
Implement the feedback I gave somewhere up there 👆 on this `mapStateToProps` as well

kimotho-njoki commented on src/redux/actions/loginAction.js:
```diff
@@ -0,0 +1,48 @@
+  type: LOGOUT_SUCCESSFUL
+});
+
+export default user => dispatch => {
```
Why is this thunk different in format from the one on line 44? 
The format for an action thunk should be 

export const user = ( ) => 
  (dispatch) => {
    ....
  }

When you import this thunk in other files, ensure you import it as user and do not rename it, to avoid confusing others when it comes to tracking down this action as the project grows.
Also, be more specific with your naming, @micahoriaso has a similar thunk, so which is which?

kimotho-njoki commented on src/redux/actions/loginAction.js:
```diff
@@ -0,0 +1,48 @@
+  error
+});
+
+export const logoutsuccessful = () => ({
```
Add comments to your action creators. Comment format

/(asterisk)(asterisk)
 (asterisk) Login successfully
 (asterisk)
 (asterisk) @param {string} message
 (asterisk) @return {object} type and payload
 (asterisk)/

Where there's asterisk please replace with the actual symbol

mikenthiwa replied:
We have implemented the above changes.

mikenthiwa replied:
@kimotho-njoki Done

kimotho-njoki commented on src/redux/actions/loginAction.js:
```diff
@@ -0,0 +1,48 @@
+      localStorage.setItem("token", message.response.token);
+      history.push(`${homepage}`);
+    })
+
```
Why this extra empty line 37?

mikenthiwa replied:
@kimotho-njoki Which line? I seem not to be seeing an empty line.

Bileonaire replied:
@kimotho-njoki We have seen and Resolved

kimotho-njoki commented on src/redux/reducers/index.js:
```diff
@@ -0,0 +1,6 @@
+
+const rootReducers = combineReducers({ loginReducer });
+
+export default rootReducers;
```
Take note @micahoriaso has a similar file but with a different name, please liaise with him and figure this out

kimotho-njoki commented on src/redux/reducers/loginReducer.js:
```diff
@@ -0,0 +1,35 @@
+  message: {},
+  error: {},
+  isAuthenticated: false
+};
```
You don't seem to have an `initialState` file. No need to create one as it has already been created by @micahoriaso . So when you rebase, you don't have to define `initialState` as you have right here. All you have to do is import `initialState` then create an object within `initialState` file to hold the contents from line 8-10 then on line 12 access `state = initialState.objectYouHaveCreated`.

kimotho-njoki commented:
@mikenthiwa @Bileonaire what's the progress with this PR? I can still see most of my feedback has not been implemented.

mikenthiwa commented:
@kimotho-njoki Currently we have implemented the changes you requested. Following your above response, we are going through the feedback to check what we have missed implementing. We have also rebased with develop, harmonised our UI with develop and pushed the changes. 
We will inform you immediately after we have finished confirming the changes.

Bileonaire commented:
@kimotho-njoki Am currently on the tests, let me sort them out.

Commit:
[Feature #159965266] Remove snapshot tests and fix action comments

pabishai commented on src/__tests__/actions/testLoginAction.js:
```diff
@@ -0,0 +1,45 @@
+const emailError = "email is required";
+const loginSuccefullMessage = "mike";
+const regURL = "http://localhost:8000/api/users/login";
+const store = mockstore({});
```
@Bileonaire and @mikenthiwa create the store in beforeEach()

micahoriaso commented:
@mikenthiwa @Bileonaire  please remove the yarn-lock file. We should have only one lock file for the application

pabishai commented on src/__tests__/actions/testLoginAction.js:
```diff
@@ -0,0 +1,48 @@
+let store;
+
+describe("action", () => {
+  beforeEach(() => {
```
this looks OK.

Bileonaire commented:
@micahoriaso Thank you for your review. The Yarn-lock has been removed.

kimotho-njoki commented on src/components/AuthButton.jsx:
```diff
@@ -0,0 +1,87 @@
+import { Logout } from "../redux/actions/loginAction";
+
+class AuthButton extends React.Component {
+  constructor(props) {
```
I can still see a bind statement, a constructor and ES5 functions. @mikenthiwa @Bileonaire

kimotho-njoki commented on src/components/login.jsx:
```diff
@@ -0,0 +1,94 @@
+import LoginForm from "./LoginForm";
+
+class LoginPage extends React.Component {
+  constructor(props) {
```
This constructor is irrelevant since the bind statement has been removed. Please refactor accordingly. You should only have the state at this point.

kimotho-njoki commented on src/redux/actions/loginAction.js:
```diff
@@ -0,0 +1,64 @@
+ * @param error
+ * @returns {{type: string, error: *}}
+ */
+
```
Why is this extra line here when I can't see it on the rest of the pages?
